File: cases/taint/go/function-field-negative/function_field_negative.go
package dataflowbench

func dfb_source() string { // DFB-SOURCE: function-field-input
	return "tainted"
}

func dfb_sink(value string) {} // DFB-SINK: function-field-sink

type Holder struct {
	Fn func(value string)
}

func invoke(target *Holder, value string) {
	target.Fn(value)
}

func run() {
	holder := &Holder{}
	otherHolder := &Holder{}
	holder.Fn = func(value string) { // DFB-WITNESS: function-field-store
		dfb_sink(value)
	}
	otherHolder.Fn = func(value string) {
		dfb_sink("clean")
	}
	invoke(otherHolder, dfb_source())
}

File: cases/taint/typescript/function-field-negative/FunctionFieldNegative.ts
function dfb_source(): string { // DFB-SOURCE: function-field-input
  return "tainted";
}

function dfb_sink(value: string): void {} // DFB-SINK: function-field-sink

interface Holder {
  fn: (value: string) => void;
}

function invoke(target: Holder, value: string): void {
  target.fn(value);
}

function run(): void {
  const holder: Holder = { fn: (value: string): void => {} };
  const otherHolder: Holder = { fn: (value: string): void => {} };
  holder.fn = (value: string): void => { // DFB-WITNESS: function-field-store
    dfb_sink(value);
  };
  otherHolder.fn = (value: string): void => {
    dfb_sink("clean");
  };
  invoke(otherHolder, dfb_source());
}
